smolyay/surrogate.py

```diff
@@ -1,546 +1,717 @@
+                                        else lookup_table[dim][ic[dim], :]
+                                    )
+                                )
+                                for dim in range(len(ic))
+                            ],
+                            axis=0,
+                        )
+                    )
+
+        # return results
+        answer.reshape(list(X.shape) + [self.num_dimensions])
+        if all(x == 1 for x in X.shape):
+            return answer.item(0)
+        else:
+            return answer
 
-    def train_from_data(self, data, linear_solver='lu'):
+    def fit(self, X, y):
         """Fit surrogate's components (basis functions) to data.
 
         Parameters
         ----------
-        data: list
+        X : array-like, UnidimensionalPointSet, or MultidimensionalPointSet of shape (n_samples, num_dimensions)
+            points that are sampled
+
+        y : array-like of shape (n_samples,)
             function at grid points.
-        linear_slover: string
-            Method of solving linear equations:
-                'lu': lower upper decomposition
-                'inv': solve for coefficients through x = A^{-1}B.
-                'lstsq': least-square solution.
+
+        Returns
+        -------
+        self : object
+            BaseSurrogate class instance
 
         Raises
         ------
         IndexError
-            Data should be of length of the grid points.
+            X must be 2D array with shape (n_samples, num_dimensions).
+        IndexError
+            y must be 1D array with shape (n_samples,).
         ValueError
-            Solver should be selected between defined methods.
+            X must lie in domain of surrogate.
         """
-        self._data = numpy.array(data, ndmin=1)
-        if self._data.shape != (len(self.grid.points),):
-            raise IndexError("Data must be same length as grid points.")
-
-        # make basis matrix
-        points, basis_functions = numpy.array(self.grid.points), self.grid.basis_functions
-        basis_matrix = numpy.zeros((len(points), len(points)))
-        for j, basis in enumerate(basis_functions):
-            if self.dimension > 1:
-                value = numpy.prod([basis[i](points[...,i]) for i in range(len(basis))],axis=0)
+        # reset constant
+        self._integration_constant = 0
+        if not self._terms_constructed_cache:
+            self._create_terms()
+            self._terms_constructed_cache = True
+
+        # get points
+        if isinstance(
+            X,
+            (
+                smolyay.samples.UnidimensionalPointSet,
+                smolyay.samples.MultidimensionalPointSet,
+            ),
+        ):
+            X = X.points
+
+        # validate data inputs
+        X = numpy.array(X,ndmin=2,copy=None)
+        self._in_domain(X)
+
+        y = numpy.asarray(y)
+        if y.shape != (X.shape[0],) and y.shape != (X.shape[0], 1):
```

Comment:
It seems like the second condition should not be allowed based on the interface and error message.
```suggestion
        if y.shape != (X.shape[0],):
```

Reply:
The reason it's like that is because if you have a 1D function, your X shape is (n_samples, 1) and most mathematical operations on a numpy array with a shape (n_samples, 1) will not change the shape. Which means if the code is that stringent about shape then you always have to use numpy.squeeze to get rid of that extra dimension even though it doesn't actually have any effect, and that can be annoying.
I'd prefer to use a slightly more lenient shape requirement just to avoid using numpy.squeeze over and over in my tests.

Reply:
Gotcha, I'm OK with allowing that in smolyay, but how does scikit-learn handle this?